### DISCUSSION TITLE
perf(debian): the dpkg analyzer reads way more files than it should

### DISCUSSION THREAD

**Discussion by paulcacheux** [aquasecurity/trivy#7081] [Bugs]
### Description

While trying to understand the performance of the dpkg/debian analyzer, I found that this analyzer basically does:
```
for every file in the FS:
   if the file is a list file (in `var/lib/dpkg/info` and `*.list` parse the info file
   in all cases parse this file as a status file
```

this means that we open and scan a lot of files, especially since parsing a status file [allocates a big buffer](https://github.com/aquasecurity/trivy/blob/1f5f34895823fae81bf521fc939bee743a50e304/pkg/fanal/analyzer/pkg/dpkg/scanner.go#L19).

### Desired Behavior

The opened and scanned files should only be related to dpkg, and in `var/lib/dpkg`

### Actual Behavior

All files of the file system are read

### Reproduction Steps

```bash
N/A
```

### Target

None

### Scanner

Vulnerability

### Output Format

None

### Mode

None

### Debug Output

```bash
N/A
```

### Operating System

debian based linux

### Version

```bash
N/A
```

### Checklist

- [ ] Run `trivy clean --all`
- [ ] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by paulcacheux** [comment]
opened https://github.com/aquasecurity/trivy/pull/7082 to improve this

**Reply by knqyf263** [reply]
```
for every file in the FS:
   if the file is a list file (in `var/lib/dpkg/info` and `*.list` parse the info file
   in all cases parse this file as a status file
```

Could you elaborate on that? [FS](https://github.com/aquasecurity/trivy/blob/1f5f34895823fae81bf521fc939bee743a50e304/pkg/fanal/analyzer/pkg/dpkg/dpkg.go#L74) includes only necessary files matching [this rule](https://github.com/aquasecurity/trivy/blob/1f5f34895823fae81bf521fc939bee743a50e304/pkg/fanal/analyzer/pkg/dpkg/dpkg.go#L291-L302). It shouldn't be a big impact on performance.

**Reply by paulcacheux** [reply]
From what I seen (I may be wrong I'm definitely not a trivy expert yet), is that this function is basically only called in tests or to enable a given analyzer, not during the walking operation. Especially this is the `required` function passed to the walk operation https://github.com/aquasecurity/trivy/blob/266d9b1f4bd2e89f1d58a16698f755a1734c644e/pkg/fanal/analyzer/pkg/dpkg/dpkg.go#L67

**Reply by paulcacheux** [reply]
Some other analyzers use their own required method to do the filtering on the other hand https://github.com/aquasecurity/trivy/blob/266d9b1f4bd2e89f1d58a16698f755a1734c644e/pkg/fanal/analyzer/language/java/gradle/lockfile.go#L51

**Reply by knqyf263** [reply]
Each `Required` is called here so unneeded files will not be stored in the virtual filesystem.
https://github.com/aquasecurity/trivy/blob/266d9b1f4bd2e89f1d58a16698f755a1734c644e/pkg/fanal/artifact/image/image.go#L274

**Reply by paulcacheux** [reply]
So yes you are right apologies for this, but if I read correctly the post analyzer FS is shared between all post analyzers right ? So if I have another post analyzer enabled that let's say needs all `*.json` files, they will all be copied in the FS and will be read by the dpkg analyzer, right ?

**Reply by knqyf263** [reply]
No, they have their own FS.
https://github.com/aquasecurity/trivy/blob/cb89fbb124a2ea7c2c49adcf27b8a2bdebc895f7/pkg/fanal/analyzer/analyzer.go#L470
